.azure-pipelines/testscripts_analyse/analyse_testcases.py

```diff
@@ -0,0 +1,124 @@
+import os
+import sys
+import uuid
+import logging
+
+from natsort import natsorted
+from datetime import datetime
+from constant import DATAPLANE_FEATURES, PR_TOPOLOGY_TYPE
+
+
+def collect_all_scripts():
+    location = sys.argv[1]
+
+    # Recursively find all files starting with "test_" and ending with ".py"
+    # Note: The full path and name of files are stored in a list named "files"
+    files = []
+    for root, dirs, file in os.walk(location):
+        for f in file:
+            if f.startswith("test_") and f.endswith(".py"):
+                files.append(os.path.join(root, f))
+    files = natsorted(files)
+
+    # Open each file and search for regex pattern
+    pattern = re.compile(r"[^@]pytest\.mark\.topology\(([^\)]*)\)")
+    test_scripts = []
+
+    for f in files:
+        # Remove prefix from file name:
+        filename = f[len(location) + 1:]
+        try:
+            with open(f, 'r') as file:
+                for line in file:
+                    match = pattern.search(line)
+                    if match:
+                        for topology in match.group(1).split(","):
+                            result = {
+                                "testscript": filename,
+                                "topology": topology.strip().strip('"').strip('\'')
+                            }
+                            test_scripts.append(result)
+        except Exception as e:
+            logging.error('Failed to load file {}, error {}'.format(f, e))
+
+    return test_scripts
+
+
+def check_PRChecker_coverd(test_scripts):
+    pr_test_scripts_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../", "pr_test_scripts.yaml")
+
+    try:
+        with open(pr_test_scripts_file) as f:
+            pr_test_scripts = yaml.safe_load(f)
+    except Exception as e:
+        logging.error('Failed to load file {}, error {}'.format(f, e))
+
+    i = 0
+    while i < len(test_scripts):
+        topo_type = test_scripts[i]["topology"]
+
+        if topo_type == "any":
+            for topology in PR_TOPOLOGY_TYPE:
+                test_scripts.append({
+                    "testscript": test_scripts[i]["testscript"],
+                    "topology": topology
+                })
+            test_scripts.remove(test_scripts[i])
+            continue
+
+        if test_scripts[i]["testscript"] in pr_test_scripts.get(topo_type, ""):
+            test_scripts[i]["covered"] = True
+        else:
+            test_scripts[i]["covered"] = False
+        i += 1
+
+
+def main():
+    test_scripts = collect_all_scripts()
+    check_PRChecker_coverd(test_scripts)
+
+    trackid = str(uuid.uuid4())
+    scantime = str(datetime.now())
+
+    for script in test_scripts:
+        script["trackid"] = trackid
+        script["scantime"] = scantime
+        if script["testscript"].split("/")[0] in DATAPLANE_FEATURES:
+            script["category"] = "data"
+        else:
+            script["category"] = "control"
+
+    return test_scripts
```

Comment:
Return an object from `main` function? It looks like the script is to return something for other code. Use `main` usually when we wish to do something by running this script.

Reply:
Yes, actually this return is no need, I will delete it